File: Python/class 1 - polymorphism.py
class Animal:
    name: str
    def __init__(self, name:str):
        self.name = name
        
    def make_sound(self):
        # do nothing
        return

class Dog(Animal):
    __breed:str
    __is_hungry:bool
    def __init__(self, name:str, breed:str):
        Animal.__init__(self, name)
        self.__breed = breed
        self.__is_hungry = True
    
    def make_sound(self):
        print(f'Woof! {self.name} is barking')
        
    def check_is_hungry(self):
        return self.__is_hungry

    def feed(self):
        if self.__is_hungry == True:
            print("Eating")
        else:
            print("The dog is not eating")
        if self.__breed == 'Pitbul':
            self.__is_hungry = True
        else:
            self.__is_hungry = False
        state = "hungry" if self.__is_hungry else "not hungry"
        print(f"The dog eat and is {state}")
        



bob:Dog = Dog("Bob", "Pitbul")
bob.feed()
bob.feed()

john:Dog = Dog("John", "Golden")
john.feed()
john.feed()

class Cat(Animal):
    def __init__(self, name:str):
        Animal.__init__(self, name)
    
    def make_sound(self):
        print(f'Meow! {self.name} is meowing')

titi:Cat = Cat("Titi")

my_list:list = [bob, titi]

for animal in my_list:
    animal.make_sound()

File: src/class 1 - polymorphism.ts
abstract class Animal {
  name: string;

  constructor(name: string) {
    this.name = name;
  }

  makeSound(): void {}
}

class Dog extends Animal {
  private breed: string;
  private isHungry: boolean;

  constructor(givenName: string, givenBreed: string) {
    super(givenName);
    this.breed = givenBreed;
    this.isHungry = true;
  }

  feed(): void {
    if (this.isHungry) {
      console.log("eating...");
    } else {
      console.log("The dog is not eating");
    }
    if (this.breed === "Pitbul") {
      this.isHungry = true;
    } else {
      this.isHungry = false;
    }
    console.log(
      `The dog eat and is ${this.isHungry ? "hungry" : "not hungry"}`
    );
  }

  checkIsHungry(): boolean {
    return this.isHungry;
  }

  makeSound(): void {
    console.log(`Woof! ${this.name} is barking`);
  }
}

const bob: Dog = new Dog("Bob", "Pitbul");
bob.feed();
bob.feed();
console.log("is bob hungry", bob.checkIsHungry());

const john: Dog = new Dog("John", "Golden Retrieve");
john.feed();
john.feed();
console.log("is john hungry", john.checkIsHungry());

class Cat extends Animal {
  constructor(name: string) {
    super(name);
  }
  makeSound(): void {
    console.log(`Meow! ${this.name} is meowing`);
  }
}

const titi: Cat = new Cat("Titi");

const myListOfAnimals: Animal[] = [bob, titi];

myListOfAnimals.forEach((animal) => animal.makeSound());
